### PR TITLE
Community: Fix sharing accounts bug & Improve Logs

### PR DESCRIPTION
**Pull request by JAVGan**
This PR fixes a bug which was causing all push items from community workflow to receive the very same sharing accounts (from the first load) instead of receiving the proper ones per account provider.
    
The fix consists on changing the `EnrichedPushItem` structure to hold now a tuple containing the push item and its related sharing accounts, which are now being retrieved from the specific `Destination` object of that push item.

It also improves the sharing accounts logs to ease the debug on how it's processing that information.
    
Refers to SPSTRAT-356

### CODE REVIEW

**Comment by JAVGan** [comment]
@lslebodn I made this fix using a "test driven development" approach, but I would like to hear your opinion on it.

My method was:

1. Download the JSON query you posted in the JIRA and store it into `tests/community_push/data/rhel-sap-community.json` (it's a HUGE json to put inside the python files)
2. Create a new test case on `tests/community_push/test_community_push.py` as follows

  ```
@pytest.mark.parametrize(
    "filename",
    ["tests/community_push/data/rhel-sap-community.json"],
    ids=["rhel-sap-community.json"],
)
@mock.patch("pubtools._marketplacesvm.tasks.community_push.CommunityVMPush.starmap")
def test_do_community_push_from_starmap_data(
    mock_starmap: mock.MagicMock,
    filename: str,
    fake_source: mock.MagicMock,
    fake_cloud_instance: mock.MagicMock,
    command_tester: CommandTester,
) -> None:
    """Test a successfull community-push from a given StArMap data."""
    # Add the custom starmap mapping
    with open(filename, 'r') as f:
        policy = json.load(f)

    mock_starmap.query_image_by_name.return_value = QueryResponse.from_json(policy)

    # Test
    command_tester.test(
        lambda: entry_point(CommunityVMPush),
        [
            "test-push",
            "--starmap-url",
            "https://starmap-example.com",
            "--credentials",
            "eyJtYXJrZXRwbGFjZV9hY2NvdW50IjogInRlc3QtbmEiLCAiYXV0aCI6eyJmb28iOiJiYXIifQo=",
            "--rhsm-url",
            "https://rhsm.com/test/api/",
            "--debug",
            "koji:https://fakekoji.com?vmi_build=ami_build",
        ],
    )
   ```

3. Perform the fixes until I got the following lines which validates the proper behavior:

```
grep "Uploading" 'tests/logs/community_push/test_community_push/test_do_community_push_from_starmap_data[rhel-sap-community.json].txt' | grep "with sharing accounts:" | sort | uniq
[    INFO] Uploading /foo/bar/sample_product_test.raw to region af-south-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-east-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-northeast-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-northeast-2 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-northeast-3 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-south-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-south-2 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-southeast-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-southeast-2 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-southeast-3 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-southeast-4 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ap-southeast-5 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ca-central-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region ca-west-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region cn-north-1 (type: hourly, ship: True) with sharing accounts: ['841519447331'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region cn-northwest-1 (type: hourly, ship: True) with sharing accounts: ['841519447331'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region eu-central-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region eu-central-2 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region eu-north-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region eu-south-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region eu-south-2 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region eu-west-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region eu-west-2 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region eu-west-3 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region il-central-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region me-central-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region me-south-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region sa-east-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region us-east-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region us-east-2 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region us-gov-east-1 (type: hourly, ship: True) with sharing accounts: ['072776908699', '505648590707'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region us-gov-west-1 (type: hourly, ship: True) with sharing accounts: ['072776908699', '505648590707'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region us-west-1 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None
[    INFO] Uploading /foo/bar/sample_product_test.raw to region us-west-2 (type: hourly, ship: True) with sharing accounts: ['279768170774', '426579533370'] and snapshot accounts: None

```

Now my question is: Should I include this test? Because the input uses the real "PROD" mappings with real billing codes and sharing_accounts used by the company so I thought it wouldn't be good to include it...

Anyways it validates that the fix works as you can see above with different sharing accounts depending on the region.

Another approach is to create a "fake" JSON file or replace the real values we have in the real input with wrong data just to upload this case... WDYT?

**Comment by jajreidy** [comment]
+1 for replacing values with fake values

**Comment by JAVGan** [comment]
> +1 for replacing values with fake values

Cool!

I've sanitized it and included on 9c634d6a1a0e6b9a49daa0f6994749b442d51e0e

@lslebodn FYI

**Review by jajreidy — APPROVED** [review]
LGTM

**Comment by lslebodn** [comment]
@JAVGan nice catch but it means the implementation of the `meta` in the MappedVMIPushItem
is very non-deterministic (dangerous)
 
https://github.com/release-engineering/pubtools-marketplacesvm/blob/0fc4b0a65c7ba99060436b2276982ec2024b1ee4/src/pubtools/_marketplacesvm/tasks/push/items/starmap.py#L60-L67

Could we remove it as part of this MR?
If it is used on other places it could contain hidden bugs as well.
If it is unused than it is safe to remove it :-)

**Comment by JAVGan** [comment]
> @JAVGan nice catch but it means the implementation of the `meta` in the MappedVMIPushItem is very non-deterministic (dangerous)
> 
> https://github.com/release-engineering/pubtools-marketplacesvm/blob/0fc4b0a65c7ba99060436b2276982ec2024b1ee4/src/pubtools/_marketplacesvm/tasks/push/items/starmap.py#L60-L67
> 
> Could we remove it as part of this MR? If it is used on other places it could contain hidden bugs as well. If it is unused than it is safe to remove it :-)

Will be done in a next PR

File: tests/logs/combined_push/test_combined_push/test_do_advisory_push.txt [hunk 1]
```diff
@@ -102,7 +151,7 @@
 [   DEBUG] Searching for product sample_product for provider awstest in rhsm
 [   DEBUG] Searching for product sample_product_HOURLY for provider awstest in rhsm
 [   DEBUG] Searching for product sample_product for provider awstest in rhsm
-[    INFO] Uploading /foo/bar/sample_product_test.raw to region us-east-1 (type: hourly, ship: True)
+[    INFO] Uploading /foo/bar/sample_product_test.raw to region us-east-1 (type: hourly, ship: True) with sharing accounts: ['0', '0', '0', '0', '0', '0', '1', '2', '1', '2', '1', '2'] and snapshot accounts: ['1', '1', '1', '1', '1', '1', '1', '2', '1', '2', '1', '2']
```
Comment (lslebodn) [inline tests/logs/combined_push/test_combined_push/test_do_advisory_push.txt:154]: @JAVGan list of characters is a bit weird in the logs; is there something wrong in mocked data?

**Comment by JAVGan** [comment]
> > @JAVGan nice catch but it means the implementation of the `meta` in the MappedVMIPushItem is very non-deterministic (dangerous)
> > https://github.com/release-engineering/pubtools-marketplacesvm/blob/0fc4b0a65c7ba99060436b2276982ec2024b1ee4/src/pubtools/_marketplacesvm/tasks/push/items/starmap.py#L60-L67
> > 
> > Could we remove it as part of this MR? If it is used on other places it could contain hidden bugs as well. If it is unused than it is safe to remove it :-)
> 
> Will be done in a next PR

This is the PR for the proposed change: https://github.com/release-engineering/pubtools-marketplacesvm/pull/39

Note that the log changes are just to include the lines from the "mapped item" as we're now processing the data for each destination individually

File: tests/logs/combined_push/test_combined_push/test_do_advisory_push.txt (hunk 1 above)
Comment (JAVGan) [reply tests/logs/combined_push/test_combined_push/test_do_advisory_push.txt:154]: I'll check it

Comment (JAVGan) [reply tests/logs/combined_push/test_combined_push/test_do_advisory_push.txt:154]: Nice catch!

The mock data was using a wrong format.

FIX: https://github.com/release-engineering/pubtools-marketplacesvm/pull/40

**Pull request merged**